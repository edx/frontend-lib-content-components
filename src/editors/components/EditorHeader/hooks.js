import React from 'react';
import * as module from './hooks';

export const hooks = {
  isEditing: () => {
    const [isEditing, setIsEditing] = React.useState(false);
    return {
      isEditing,
      startEditing: () => setIsEditing(true),
      stopEditing: () => setIsEditing(false),
    };
  },
  localTitle: ({
    setBlockTitle,
    stopEditing,
    typeHeader,
  }) => {
    const [localTitle, setLocalTitle] = React.useState(typeHeader);
    return {
      updateTitle: () => {
        setBlockTitle(localTitle);
        stopEditing();
      },
      handleChange: (e) => setLocalTitle(e.target.value),
      localTitle,
    };
  },
  handleKeyDown: ({ stopEditing, editorRef }) => (e) => {
    if (e.key === 'Enter') {
      stopEditing();
    }
    if (e.key === 'Tab' && editorRef) {
      e.preventDefault();
      editorRef.current.focus();
    }
  },
};

/* eslint-disable import/prefer-default-export */
export const localTitleHooks = ({
  editorRef,
  setBlockTitle,
  typeHeader,
}) => {
  const { isEditing, startEditing, stopEditing } = module.hooks.isEditing();
  const { localTitle, handleChange, updateTitle } = module.hooks.localTitle({
    setBlockTitle,
    stopEditing,
    typeHeader,
  });
  return {
    isEditing,
    startEditing,
    stopEditing,

    localTitle,
    updateTitle,
    handleChange,

    inputRef: React.createRef(),
    handleKeyDown: module.hooks.handleKeyDown({ stopEditing, editorRef }),
  };
};
