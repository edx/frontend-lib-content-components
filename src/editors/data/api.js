import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { ActionStates, normalizeContent } from './constants';

async function getAsync(updateContext, params) {
  try {
    updateContext.setLoading(ActionStates.IN_PROGRESS);
    const result = await getAuthenticatedHttpClient().get(...params);
    updateContext.setValue(result);
  } catch (e) {
    updateContext.setError(e);
  } finally {
    updateContext.setLoading(ActionStates.FINISHED);
  }
}

async function saveAsync(updateContext, params) {
  try {
    const result = await getAuthenticatedHttpClient().post(...params);
    updateContext.setResponse(result);
  } catch (e) {
    updateContext.setResponse(e);
  } finally {
    updateContext.setInProgress(ActionStates.FINISHED);
  }
}

export async function fetchBlockById(updateContext, blockId, studioEndpointUrl) {
  const url = `${studioEndpointUrl}/xblock/${blockId}`;
  getAsync(updateContext, [url]);
}

export async function fetchUnitById(updateContext, blockId, studioEndpointUrl) {
  const url = `${studioEndpointUrl}/xblock/${blockId}?fields=ancestorInfo`;
  getAsync(updateContext, [url]);
}

export async function saveBlock(blockId, blockType, courseId, studioEndpointUrl, content, updateContext) {
  const normalizedContent = normalizeContent(blockType, content, blockId, courseId);
  const url = `${studioEndpointUrl}/xblock/${blockId}`;
  const params = [url, normalizedContent];
  saveAsync(updateContext, params);
}
