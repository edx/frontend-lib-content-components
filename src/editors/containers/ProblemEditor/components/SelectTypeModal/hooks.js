import {
  useState,
} from 'react';
import { StrictDict } from '../../utils';

import * as module from './hooks';

export const state = StrictDict({
  selected: (val) => useState(val),
});

export default { state };
