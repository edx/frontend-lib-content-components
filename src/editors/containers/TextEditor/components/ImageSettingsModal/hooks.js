import React from 'react';

import { StrictDict } from '../../../../utils';
import * as module from './hooks';

// Simple wrappers for useState to allow easy mocking for tests.
export const state = {
  dimensions: (val) => React.useState(val),
  locked: (val) => React.useState(val),
  local: (val) => React.useState(val),
  lockInitialized: (val) => React.useState(val),
  altText: (val) => React.useState(val),
  isDecorative: (val) => React.useState(val),
};

export const dimKeys = StrictDict({
  height: 'height',
  width: 'width',
});

/**
 * findGcd(numerator, denominator)
 * Find the greatest common denominator of a ratio or fraction.
 * @param {number} numerator - ratio numerator
 * @param {number} denominator - ratio denominator
 * @return {number} - ratio greatest common denominator
 */
export const findGcd = (a, b) => (b ? findGcd(b, a % b) : a);
const checkEqual = (d1, d2) => (d1.height === d2.height && d1.width === d2.width);

/**
 * getValidDimensions({ dimensions, local, locked })
 * Find valid ending dimensions based on start state, request, and lock state
 * @param {obj} dimensions - current stored dimensions
 * @param {obj} local - local (active) dimensions in the inputs
 * @param {obj} locked - locked dimensions
 * @return {obj} - output dimensions after move ({ height, width })
 */
export const getValidDimensions = ({
  dimensions,
  local,
  locked,
}) => {
  const out = {};
  let iter;
  const { minInc } = locked;
  const isMin = dimensions.height === minInc.height;

  const keys = (local.height !== dimensions.height)
    ? { changed: dimKeys.height, other: dimKeys.width }
    : { changed: dimKeys.width, other: dimKeys.height };

  const direction = local[keys.changed] > dimensions[keys.changed] ? 1 : -1;

  // don't move down if already at minimum size
  if (direction < 0 && isMin) { return dimensions; }
  // find closest valid iteration of the changed field
  iter = Math.max(Math.round(local[keys.changed] / minInc[keys.changed]), 1);
  // if closest valid iteration is current iteration, move one iteration in the change direction
  if (iter === (dimensions[keys.changed] / minInc[keys.changed])) { iter += direction; }

  out[keys.changed] = iter * minInc[keys.changed];
  out[keys.other] = out[keys.changed] * (locked[keys.other] / locked[keys.changed]);

  return out;
};

/**
 * newDimensions({ dimensions, local, locked })
 * Returns the local dimensions if unlocked or unchanged, and otherwise returns new valid
 * dimensions.
 * @param {obj} dimensions - current stored dimensions
 * @param {obj} local - local (active) dimensions in the inputs
 * @param {obj} locked - locked dimensions
 * @return {obj} - output dimensions after attempted move ({ height, width })
 */
export const newDimensions = ({ dimensions, local, locked }) => (
  (!locked || checkEqual(local, dimensions))
    ? local
    : module.getValidDimensions({ dimensions, local, locked })
);

/**
 * lockDimensions({ dimensions, lockInitialized, setLocked })
 * Lock dimensions if lock initialized.  Store minimum valid increment on lock so
 * that we don't have re-compute.
 * @param {obj} dimensions - current stored dimensions
 * @param {bool} lockInitialized - has the lock state initialized?
 * @param {func} setLocked - set lock state
 */
export const lockDimensions = ({ dimensions, lockInitialized, setLocked }) => {
  if (!lockInitialized) { return; }

  // find minimum viable increment
  let gcd = findGcd(dimensions.width, dimensions.height);
  if ([dimensions.width, dimensions.height].some(v => !Number.isInteger(v / gcd))) {
    gcd = 1;
  }
  const minInc = { width: dimensions.width / gcd, height: dimensions.height / gcd, gcd };
  setLocked({ ...dimensions, minInc });
};

/**
 * dimensionLockHooks({ dimensions })
 * Returns a set of hooks pertaining to the dimension locks.
 * Locks the dimensions initially, on lock initialization.
 * @param {obj} dimensions - current stored dimensions
 * @return {obj} - dimension lock hooks
 *   {func} initializeLock - enable the lock mechanism
 *   {obj} locked - current locked state
 *   {func} lock - lock the current dimensions
 *   {func} unlock - unlock the dimensions
 */
export const dimensionLockHooks = ({ dimensions }) => {
  const [locked, setLocked] = module.state.locked(null);
  const [lockInitialized, setLockInitialized] = module.state.lockInitialized(null);
  const lock = () => module.lockDimensions({ lockInitialized, dimensions, setLocked });

  React.useEffect(lock, [lockInitialized]);

  return {
    initializeLock: () => setLockInitialized(true),
    locked,
    lock,
    unlock: () => setLocked(null),
  };
};

/**
 * dimensionHooks()
 * Returns an object of dimension-focused react hooks.
 * @return {obj} - dimension hooks
 *   {func} onImgLoad - initializes image dimension fields
 *     @param {object} selection - selected image object with possible override dimensions.
 *     @return {callback} - image load event callback that loads dimensions.
 *   {object} locked - current locked state
 *   {func} lock - lock current dimensions
 *   {func} unlock - unlock dimensions
 *   {object} value - current dimension values
 *   {func} setHeight - set height
 *     @param {string} - new height string
 *   {func} setWidth - set width
 *     @param {string} - new width string
 *   {func} updateDimensions - set dimensions based on state
 */
export const dimensionHooks = () => {
  const [dimensions, setDimensions] = module.state.dimensions(null);
  const [local, setLocal] = module.state.local(null);
  const setAll = (value) => {
    setDimensions(value);
    setLocal(value);
  };
  const {
    initializeLock,
    lock,
    locked,
    unlock,
  } = module.dimensionLockHooks({ dimensions });
  return {
    onImgLoad: (selection) => ({ target: img }) => {
      setAll({
        height: selection.height || img.naturalHeight,
        width: selection.width || img.naturalWidth,
      });
      initializeLock();
    },
    locked,
    lock,
    unlock,
    value: local,
    setHeight: (height) => setLocal({ ...local, height: parseInt(height, 10) }),
    setWidth: (width) => setLocal({ ...local, width: parseInt(width, 10) }),
    updateDimensions: () => setAll(module.newDimensions({ dimensions, local, locked })),
  };
};

/**
 * altTextHooks(savedText)
 * Returns a set of react hooks focused around alt text
 * @return {obj} - alt text hooks
 *   {string} value - alt text value
 *   {func} setValue - set alt test value
 *     @param {string} - new alt text
 *   {bool} isDecorative - is the image decorative?
 *   {func} setIsDecorative - set isDecorative field
 *     @param {bool} isDecorative
 */
export const altTextHooks = (savedText) => {
  const [value, setValue] = module.state.altText(savedText || '');
  const [isDecorative, setIsDecorative] = module.state.isDecorative(false);
  return {
    value,
    setValue,
    isDecorative,
    setIsDecorative,
  };
};

/**
 * onInputChange(handleValue)
 * Simple event handler forwarding the event target value to a given callback
 * @param {func} handleValue - event value handler
 * @return {func} - evt callback that will call handleValue with the event target value.
 */
export const onInputChange = (handleValue) => (e) => handleValue(e.target.value);

/**
 * onCheckboxChange(handleValue)
 * Simple event handler forwarding the event target checked prop to a given callback
 * @param {func} handleValue - event value handler
 * @return {func} - evt callback that will call handleValue with the event target checked prop.
 */
export const onCheckboxChange = (handleValue) => (e) => handleValue(e.target.checked);

/**
 * onSave({ altText, dimensions, isDecorative, saveToEditor })
 * Handle saving the image context to the text editor
 * @param {string} altText - image alt text
 * @param {object} dimension - image dimensions ({ width, height })
 * @param {bool} isDecorative - is the image decorative?
 * @param {func} saveToEditor - save method for submitting image settings.
 */
export const onSaveClick = ({
  altText,
  dimensions,
  isDecorative,
  saveToEditor,
}) => () => saveToEditor({
  altText,
  dimensions,
  isDecorative,
});

/**
 * isSaveDisabled(altText)
 * Returns true the save button should be disabled (altText is missing and not decorative)
 * @param {object} altText - altText hook object
 *   {bool} isDecorative - is the image decorative?
 *   {string} value - alt text value
 * @return {bool} - should the save button be disabled?
 */
export const isSaveDisabled = (altText) => !altText.isDecorative && (altText.value === '');

export default {
  altText: altTextHooks,
  dimensions: dimensionHooks,
  isSaveDisabled,
  onCheckboxChange,
  onInputChange,
  onSaveClick,
};
