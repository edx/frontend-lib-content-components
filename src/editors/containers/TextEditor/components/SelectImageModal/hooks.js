import React from 'react';
import * as module from './hooks';
import { sortFunctions, sortKeys } from './utils';

export const state = {
  images: (val) => React.useState(val),
  highlighted: (val) => React.useState(val),
  searchString: (val) => React.useState(val),
  sortBy: (val) => React.useState(val),
};

export const searchAndSortHooks = () => {
  const [searchString, setSearchString] = module.state.searchString('');
  const [sortBy, setSortBy] = module.state.sortBy(sortKeys.dateNewest);
  return {
    searchString,
    onSearchChange: e => setSearchString(e.target.value),
    clearSearchString: () => setSearchString(''),
    sortBy,
    onSortClick: key => () => setSortBy(key),
  };
};

export const filteredList = ({ searchString, imageList }) => imageList.filter(
  ({ displayName }) => displayName.toLowerCase().includes(searchString.toLowerCase()),
);

export const displayList = ({ sortBy, searchString, images }) => module.filteredList({
  searchString,
  imageList: Object.values(images),
}).sort(sortFunctions[sortBy in sortKeys ? sortKeys[sortBy] : sortKeys.dateNewest]);

export const imgListHooks = ({
  fetchImages,
  setSelection,
  searchSortProps,
}) => {
  const [images, setImages] = module.state.images({});
  const [highlighted, setHighlighted] = module.state.highlighted(null);

  React.useEffect(() => {
    fetchImages({ onSuccess: setImages });
  }, []);

  return {
    images,
    // highlight by id
    selectBtnProps: {
      disabled: !highlighted,
      onClick: () => setSelection(images[highlighted]),
    },
    galleryProps: {
      highlighted,
      onHighlightChange: e => setHighlighted(e.target.value),
      displayList: module.displayList({ ...searchSortProps, images }),
    },
  };
};

export const fileInputHooks = ({ uploadImage }) => {
  const ref = React.useRef();
  const click = () => ref.current.click();
  const resetFile = () => { ref.current.value = ''; };
  const addFile = (e) => uploadImage({
    file: e.target.files[0],
    resetFile,
  });
  return {
    click,
    addFile,
    ref,
  };
};

export const imgHooks = ({ fetchImages, uploadImage, setSelection }) => {
  const searchSortProps = module.searchAndSortHooks();
  const imgList = module.imgListHooks({ fetchImages, setSelection, searchSortProps });
  const fileInput = module.fileInputHooks({ uploadImage });
  const { selectBtnProps, galleryProps } = imgList;

  return {
    fileInput,
    galleryProps,
    selectBtnProps,
    searchSortProps,
  };
};

export default {
  imgHooks,
};
