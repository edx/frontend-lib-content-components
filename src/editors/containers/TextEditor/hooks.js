import {
  useRef, useEffect, useCallback, useState,
} from 'react';

import { StrictDict } from '../../utils';
import tinyMCE from '../../data/constants/tinyMCE';
import pluginConfig from './pluginConfig';
import * as appHooks from '../../hooks';
import * as module from './hooks';

export const { nullMethod, navigateCallback, navigateTo } = appHooks;

export const state = StrictDict({
  isModalOpen: (val) => useState(val),
  imageSelection: (val) => useState(val),
  refReady: (val) => useState(val),
});

/**
 * Create A function which takes in a tinyMCE editor instance and adds custom
 * features to the toolbars and UI.
 *
 * @param {openModal} func - a function which opens the custom image modal.
 * @param {setImage} func - a function which sets the current image selected in the image modal.
 * @return {func} - takes in a tinyMCE editor instance and adds custom
 * features to the toolbars and UI.
 */

export const setupCustomBehavior = ({ openModal, setImage }) => (editor) => {
  // image upload button
  editor.ui.registry.addButton(tinyMCE.buttons.imageUploadButton, {
    icon: 'image',
    onAction: openModal,
  });
  // editing an existing image
  editor.ui.registry.addButton(tinyMCE.buttons.editImageSettings, {
    icon: 'image',
    onAction: module.openModalWithSelectedImage({ editor, setImage, openModal }),
  });
  // overriding the code plugin's icon with 'HTML' text
  const openCodeEditor = () => editor.execCommand('mceCodeEditor');
  editor.ui.registry.addButton(tinyMCE.buttons.code, {
    text: 'HTML',
    tooltip: 'Source code',
    onAction: openCodeEditor,
  });
};

/**
 * Setup tinyMCE editor with chosen inputs
 *
 * @param {setEditorRef} func - a function which sets the higher-level reference to the editor.
 * @param {blockValue} object - the xblock object which contains the html.
 * @param {openModal} func - a function which opens the custom image modal.
 * @param {setSelection} func - a function which sets the current image selected in the image modal.
 * @param {initializeEditor} func - a function which tells the higher level data abstraction
 * the editor is ready to be displayed.
 */

export const editorConfig = ({
  setEditorRef,
  blockValue,
  openModal,
  initializeEditor,
  setSelection,
}) => ({
  onInit: (evt, editor) => {
    setEditorRef(editor);
    initializeEditor();
  },
  initialValue: blockValue ? blockValue.data.data : '',
  init: {

    setup: module.setupCustomBehavior({ openModal, setImage: setSelection }),
    plugins: pluginConfig.plugins,
    imagetools_toolbar: pluginConfig.imageToolbar,
    toolbar: pluginConfig.toolbar,
    contextmenu: 'link table',
    ...pluginConfig.config,
    valid_elements: '*[*]',
  },
});

/**
 * A stateful toggle for opening/closing a modal.
 * @return {object} - An with toggle open/closed functions and a state reference.
 *
 * usage:
 *   const {isModalOpen: isOpen, openModal, closeModal} = modalToggle();
 *   openModal();
 *   closeModal();
 */

export const modalToggle = () => {
  const [isOpen, setIsOpen] = module.state.isModalOpen(false);
  return {
    isOpen,
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  };
};

/**
 * Creates a function which opens the image modal with the selected image when called.
 *
 * @param {editor} object - a tinyMCE editor
 * @param {openModal} func - a function which opens the custom image modal.
 * @param {setImage} func - a function which sets the current image selected in the image modal.
 *
 * @return {func} - An with toggle open/closed functions and a state reference.
 *
 * usage:
 *   const openModal= openModalWithSelectedImage ({editor: currentMCE, setImage, openModal})
 *   openModal();
 */

export const openModalWithSelectedImage = ({ editor, setImage, openModal }) => () => {
  const imgHTML = editor.selection.getNode();
  setImage({
    externalUrl: imgHTML.src,
    altText: imgHTML.alt,
    width: imgHTML.width,
    height: imgHTML.height,
  });
  openModal();
};

export const prepareEditorRef = () => {
  const editorRef = useRef(null);
  const setEditorRef = useCallback((ref) => {
    editorRef.current = ref;
  }, []);
  const [refReady, setRefReady] = module.state.refReady(false);
  useEffect(() => setRefReady(true), []);
  return { editorRef, refReady, setEditorRef };
};

/**
 * Creates a function which returns the HTML content of the tinyMCE Editor
 *
 * @param {editorRef} object - a reference to the tinyMCE editor instance in the DOM
 * @return {func} - a function which returns the content of the editor.
 */

export const getContent = ({ editorRef }) => () => editorRef.current?.getContent();

export const selectedImage = (val) => {
  const [selection, setSelection] = module.state.imageSelection(val);
  return {
    clearSelection: () => setSelection(null),
    selection,
    setSelection,
  };
};
