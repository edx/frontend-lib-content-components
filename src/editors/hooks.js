import {
  useRef, useEffect, useCallback, useState,
} from 'react';
import * as module from './hooks';

export const initializeApp = ({ initialize, data }) => useEffect(() => initialize(data), []);

export const prepareEditorRef = () => {
  const editorRef = useRef(null);
  const setEditorRef = useCallback((ref) => {
    editorRef.current = ref;
  }, []);
  const [refReady, setRefReady] = useState(false);
  useEffect(() => setRefReady(true), []);
  return { editorRef, refReady, setEditorRef };
};

export const navigateTo = (destination) => {
  window.location.assign(destination);
};

export const navigateCallback = (destination) => () => navigateTo(destination);

export const saveTextBlock = ({
  editorRef,
  returnUrl,
  saveBlock,
}) => {
  console.log(editorRef);
  saveBlock({
    returnToUnit: module.navigateCallback(returnUrl),
    content: editorRef.current.getContent(),
  });
};

// for toast onClose to avoid console warnings
export const nullMethod = () => {};
