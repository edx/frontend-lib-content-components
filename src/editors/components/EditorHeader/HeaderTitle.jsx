import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Icon, IconButton } from '@edx/paragon';
import { Edit } from '@edx/paragon/icons';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';

import { actions, selectors } from '../../data/redux';
import { localTitleHooks } from './hooks';
import messages from './messages';
import EditableHeader from './EditableHeader';

export const HeaderTitle = ({
  editorRef,
  intl,
  isInitialized,
  setBlockTitle,
  typeHeader,
}) => {
  if (!isInitialized) { return intl.formatMessage(messages.loading); }
  const {
    inputRef,
    isEditing,
    handleChange,
    handleKeyDown,
    localTitle,
    startEditing,
    updateTitle,
  } = localTitleHooks({
    editorRef,
    setBlockTitle,
    typeHeader,
  });

  if (isEditing) {
    return (
      <EditableHeader
        {...{
          inputRef,
          handleChange,
          handleKeyDown,
          localTitle,
          updateTitle,
        }}
      />
    );
  }
  return (
    <div className="d-flex">
      <div style={{ lineHeight: '1.5', paddingRight: '.25em' }}>
        {localTitle}
      </div>
      <IconButton
        alt={intl.formatMessage(messages.editTitleLabel)}
        className="mr-2"
        iconAs={Icon}
        onClick={startEditing}
        size="sm"
        src={Edit}
      />
    </div>
  );
};
HeaderTitle.defaultProps = {
  editorRef: null,
};
HeaderTitle.propTypes = {
  editorRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  // injected
  intl: intlShape.isRequired,
  // redux
  isInitialized: PropTypes.bool.isRequired,
  setBlockTitle: PropTypes.func.isRequired,
  typeHeader: PropTypes.string.isRequired,
};

export const mapStateToProps = (state) => ({
  typeHeader: selectors.app.typeHeader(state),
  isInitialized: selectors.app.isInitialized(state),
});

export const mapDispatchToProps = {
  setBlockTitle: actions.app.setBlockTitle,
};

export default injectIntl(connect(mapStateToProps, mapDispatchToProps)(HeaderTitle));
