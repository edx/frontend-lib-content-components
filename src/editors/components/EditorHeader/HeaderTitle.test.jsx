import React from 'react';
import { shallow } from 'enzyme';

import * as module from './HeaderTitle';
import { actions, selectors } from '../../data/redux';
import { localTitleHooks } from './hooks';

jest.mock('../../data/redux', () => ({
  actions: {
    app: {
      setBlockTitle: jest.fn().mockName('actions.app.setBlockTitle'),
    },
  },
  selectors: {
    app: {
      typeHeader: jest.fn(state => ({ typeHeader: state })),
      isInitialized: jest.fn(state => ({ studioEndpointUrl: state })),
    },
  },
}));

jest.mock('./hooks', () => ({
  localTitleHooks: jest.fn(),
}));

describe('HeaderTitle', () => {
  const props = {
    editorRef: jest.fn().mockName('args.editorRef'),
    isInitialized: false,
    setBlockTitle: jest.fn().mockName('args.setBlockTitle'),
    typeHeader: 'html',
  };
  const localTitleHooksProps = {
    inputRef: jest.fn().mockName('localTitleHooks.inputRef'),
    isEditing: false,
    handleChange: jest.fn().mockName('localTitleHooks.handleChange'),
    handleKeyDown: jest.fn().mockName('localTitleHooks.handleKeyDown'),
    localTitle: 'TeST LocALtitLE',
    startEditing: jest.fn().mockName('localTitleHooks.startEditing'),
    updateTitle: jest.fn().mockName('localTitleHooks.updateTitle'),
  };

  describe('behavior', () => {
    it(' calls localTitleHooks with initialization args', () => {
      localTitleHooks.mockReturnValue(localTitleHooksProps);
      shallow(<module.HeaderTitle {...props} isInitialized />);
      expect(localTitleHooks).toHaveBeenCalledWith({
        editorRef: props.editorRef,
        setBlockTitle: props.setBlockTitle,
        typeHeader: props.typeHeader,
      });
    });
  });

  describe('snapshots', () => {
    test('not initialized', () => {
      expect(shallow(<module.HeaderTitle {...props} />)).toMatchSnapshot();
    });
    test('initialized', () => {
      localTitleHooks.mockReturnValue(hookProps);
      expect(shallow(<module.HeaderTitle {...props} isInitialized />)).toMatchSnapshot();
    });
    test('editing', () => {
      localTitleHooks.mockReturnValue({ ...hookProps, isEditing: true });
      expect(shallow(<module.HeaderTitle {...props} isInitialized />)).toMatchSnapshot();
    });
  });

  describe('mapStateToProps', () => {
    const testState = { T: 'esting', S: 'tate' };
    test('typeHeader from app.typeHeader', () => {
      expect(module.mapStateToProps(testState).typeHeader)
        .toEqual(selectors.app.typeHeader(testState));
    });
    test('isInitialized from app.isInitialized', () => {
      expect(module.mapStateToProps(testState).isInitialized)
        .toEqual(selectors.app.isInitialized(testState));
    });
  });

  describe('mapDispatchToProps', () => {
    test('setBlockTitle from actions.app.setBlockTitle', () => {
      expect(module.mapDispatchToProps.setBlockTitle).toEqual(actions.app.setBlockTitle);
    });
  });
});
