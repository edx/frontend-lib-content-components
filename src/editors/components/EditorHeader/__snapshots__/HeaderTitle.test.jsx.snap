// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderTitle snapshots editing 1`] = `
<EditableHeader
  handleChange={[MockFunction localTitleHooks.handleChange]}
  handleKeyDown={[MockFunction localTitleHooks.handleKeyDown]}
  inputRef={[MockFunction localTitleHooks.inputRef]}
  localTitle="TeST LocALtitLE"
  updateTitle={[MockFunction localTitleHooks.updateTitle]}
/>
`;

exports[`HeaderTitle snapshots initialized 1`] = `
<div
  className="d-flex"
>
  <div
    style={
      Object {
        "lineHeight": "1.5",
        "paddingRight": ".25em",
      }
    }
  >
    TeST LocALtitLE
  </div>
  <IconButton
    alt="Edit Title"
    className="mr-2"
    iconAs="Icon"
    onClick={[MockFunction localTitleHooks.startEditing]}
    size="sm"
    src={[MockFunction icons.Edit]}
  />
</div>
`;

exports[`HeaderTitle snapshots not initialized 1`] = `"Loading..."`;
